### PR TITLE
Clear timers on warp, make level callbacks better

### PR DESCRIPTION
**Pull request by Dregu**
After adding warps the timers were not cleared cause the screen never changes. Also the checks in onlevel were always a bit janky and I think they only worked 99.5%.

I think state.loading should work for both. Does this make more sense?

Also returning false from interval to clear it sounds neat as in #78

### CODE REVIEW

File: src/game_api/script/lua_backend.cpp [hunk 1]
```diff
@@ -276,8 +279,13 @@ bool LuaBackend::update()
             {
                 if (now >= cb->lastRan + cb->interval)
                 {
-                    handle_function(cb->func);
+                    std::optional<bool> clear_me = handle_function_with_return<bool>(cb->func);
                     cb->lastRan = now;
+                    if (clear_me && clear_me.value() == false)
```
Comment (Malacath-92) [inline src/game_api/script/lua_backend.cpp:284]: This condition could just be `if (clear_me.value_or(false))`

File: src/game_api/script/lua_backend.cpp [hunk 2]
```diff
@@ -176,10 +176,13 @@ bool LuaBackend::update()
         std::vector<Player*> players = get_players();
         lua["players"] = std::vector<Player*>(players);
 
-        if (g_state->screen != state.screen && g_state->screen_last != 5)
+        if (g_state->loading == 1 && g_state->loading != state.loading)
```
Comment (Malacath-92) [inline src/game_api/script/lua_backend.cpp:179]: Don't we want the `screen != options && screen_last != options` check here?

File: src/game_api/script/lua_backend.cpp [hunk 3]
```diff
@@ -176,10 +176,13 @@ bool LuaBackend::update()
...
         {
             level_timers.clear();
             script_input.clear();
+        }
+        if (g_state->screen != state.screen && g_state->screen_last != 5)
```
Comment (Malacath-92) [inline src/game_api/script/lua_backend.cpp:184]: Instead here, why do we only check moving from options and not moving to options? Or check neither. As it's only a callback users might be interested in moving to the options screen.

File: src/game_api/script/lua_backend.cpp [hunk 4]
```diff
@@ -355,8 +363,7 @@ bool LuaBackend::update()
                 }
                 case ON::START:
                 {
-                    if (g_state->screen == (int)ON::LEVEL && g_state->level_count == 0 && !players.empty() &&
-                        state.player != players.at(0))
+                    if (g_state->screen == (int)ON::LEVEL && g_state->level_count == 0 && state.loading != g_state->loading && g_state->loading == 3)
```
Comment (Malacath-92) [inline src/game_api/script/lua_backend.cpp:366]: This one also runs when coming back from options right now. So I guess it needs a `last_screen != options` or `state.time_level > g_state->time_level`

File: src/game_api/script/lua_backend.cpp [hunk 5]
```diff
@@ -311,12 +319,12 @@ bool LuaBackend::update()
...
         for (auto& [id, callback] : callbacks)
         {
-            if ((ON)g_state->screen == callback.screen && g_state->screen != state.screen && g_state->screen_last != 5) // game screens
+            if ((ON)g_state->screen == callback.screen && g_state->screen != state.screen && g_state->screen_last != (int)ON::OPTIONS) // game screens
             {
                 handle_function(callback.func);
                 callback.lastRan = now;
             }
-            else if (callback.screen == ON::LEVEL && g_state->screen == (int)ON::LEVEL && g_state->screen_last != (int)ON::OPTIONS && !players.empty() && (state.player != players.at(0) || ((g_state->quest_flags & 1) == 0 && state.reset > 0)))
+            else if (callback.screen == ON::LEVEL && g_state->screen == (int)ON::LEVEL && state.loading != g_state->loading && g_state->loading == 3)
```
Comment (Malacath-92) [inline src/game_api/script/lua_backend.cpp:327]: This actually also still runs when coming back from options

File: src/game_api/script/lua_backend.cpp (hunk 3 above)
Comment (Dregu) [reply src/game_api/script/lua_backend.cpp:184]: My brain really went backwards here didn't it if I put the options check in the only place it wasn't needed.

**Comment by Dregu** [comment]
I guess my logic was relying on that visiting options doesn't use state.loading but leaving to main menu somehow does.

**Comment by Malacath-92** [comment]
Added a fix for #138

**Review by Malacath-92 — APPROVED** [review]

**Pull request merged**